vm/src/eval.rs

```diff
@@ -1,27 +1,32 @@
-//! Evaluation for Nexus VM programs.
+//! A Virtual Machine for RISC-V
 
-use num_traits::FromPrimitive;
+use crate::{
+    error::*,
+    memory::Memory,
+    rv32::{parse::*, *},
+    syscalls::Syscalls,
+};
 
-use crate::error::{NexusVMError::InvalidInstruction, Result};
-use crate::instructions::{Inst, Opcode, Opcode::*, Width};
-use crate::memory::Memory;
-use crate::syscalls::Syscalls;
+use std::collections::HashSet;
 
-/// State of a running Nexus VM program.
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use serde::{Deserialize, Serialize};
+
+/// virtual machine state
 #[derive(Default)]
 pub struct NexusVM<M: Memory> {
-    /// Current program counter.
-    pub pc: u32,
-    /// Register file.
-    pub regs: [u32; 32],
-    /// Most recent instruction.
-    pub inst: Inst,
-    /// Result of most recent instruction.
-    pub Z: u32,
+    /// ISA registers
+    pub regs: Regs,
     /// Syscall implementation
     pub syscalls: Syscalls,
+    /// current instruction
+    pub inst: Inst,
+    /// internal result register
+    pub Z: u32,
+    /// used instruction sets
+    pub instruction_sets: HashSet<InstructionSet>,
```

Comment:
nit: Is this needed now? I see where it's initialized in the `new()` constructor and inserted into in `eval_inst()`, but haven't yet found any reads.

I'm not opposed to including it if it's going to help, but also wonder if we may find a more robust abstraction later. In this context, the `Inst` used for the `inst` already specifies a sort of instruction set that the `NexusVM` should be able to handle.

Reply:
This will get used as soon as I rebase on https://github.com/nexus-xyz/nexus-zkvm/pull/159, in order to block ecalls with Jolt until I've added support for it.